listing/listing.go

```diff
@@ -0,0 +1,195 @@
+package listing
+
+import (
+	"context"
+	"errors"
+)
+
+// Iterator[T] is an iterator over items of type T. It is similar to
+// java.util.Iterator. It is not thread-safe.
+type Iterator[T any] interface {
+	// HasNext returns true if there are more items to iterate over. HasNext
+	// will also return true if the iterator needs to fetch the next page of
+	// items from the underlying source and the request fails, even if there
+	// are no more items to iterate over. In this case, Next will return the
+	// error.
+	HasNext(context.Context) bool
+
+	// Next returns the next item in the iterator. If there are no more items
+	// to iterate over, it returns ErrNoMoreItems. If there was an error
+	// fetching the next page of items, it returns that error. Once Next
+	// returns ErrNoMoreItems or an error, it will continue to return that
+	// error.
+	Next(context.Context) (T, error)
+}
+
+// ToSlice returns all items from the iterator as a slice. If there was an
+// error fetching items at any time, it returns that error.
+func ToSlice[T any](ctx context.Context, it Iterator[T]) ([]T, error) {
+	var items []T
+	for it.HasNext(ctx) {
+		item, err := it.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+		items = append(items, item)
+	}
+	return items, nil
+}
+
+// Use struct{} for Req for iterators that don't need any request structure.
+type iteratorImpl[Req, Resp, T any] struct {
+	// nextReq is the request to be used to fetch the next page. If nil, then
+	// there is no next page to fetch.
+	nextReq *Req
+
+	// getNextPage fetches the next page of items, returning the deserialized
+	// response and error.
+	getNextPage func(context.Context, Req) (Resp, error)
+
+	// getItems selects the items being iterated over from the response.
+	getItems func(Resp) []T
+
+	// getNextReq is used to get the next request to be used in the next page.
+	// If nil, then there is no next page to fetch.
+	getNextReq func(Resp) *Req
+
+	// currentPage is the current page of items.
+	currentPage []T
+
+	// currentPageIdx is the index of the next item from currentPage to return.
+	currentPageIdx int
+
+	// lastErr is the last error returned by getNextPage.
+	lastErr error
+}
+
+var ErrNoMoreItems = errors.New("no more items")
+
+// Returns a new iterator. The iterator will fetch the next page of items
+// lazily, when needed. nextReq is the request to be used to fetch the initial
+// page. If nil, then no page will be fetched. getNextPage fetches the next
+// page of items, returning the deserialized response and error. getItems
+// selects the items being iterated over from the response. getNextReq is used
+// to get the next request to be used in the next page. If the returned value
+// is nil, then there is no next page to fetch.
+func NewIterator[Req, Resp, T any](
+	nextReq *Req,
+	getNextPage func(context.Context, Req) (Resp, error),
+	getItems func(Resp) []T,
+	getNextReq func(Resp) *Req,
+) *iteratorImpl[Req, Resp, T] {
```

Comment:
Why not return `Iterator[T]`, this being a public function?

Reply:
I think it's kind of up to users if they want to define interfaces, right? We define the interface so that we can provide this to NewDedupeIterator, but if we add more public methods to this base type, users should in theory be able to use them.

Reply:
It's a unexported type though, so for documentation sake I guess it would be better to document it as returning an `Iterator[T]`. As is, it will show up as `unexported type`.

Reply:
Actually, I was thinking about this a bit more, and I think we may want to export this type as well and have the List implementations return the actual concrete type. None of the fields will be exported, so users won't be able to modify iterator state, but e.g. if we define more methods on either of those structures that don't make sense for the overall iterator (e.g. imagine a "GetCached" on DedupeIteratorImpl that returns the cached element or nil), we couldn't expose that without changing the Iterator interface or using the concrete type.